### ISSUE TITLE
[BUG] Beam search produces wrong results when num_beams is not a power of two 

### ISSUE THREAD

**Issue by merrymercy** [alpa-projects/alpa#580]
The current beam search can produce correct results when num_beams is a power of two. But the results are wrong otherwise. 

To reproduce:
1. Comment out this line.
  https://github.com/alpa-projects/alpa/blob/b8c4329f403621913e84a202086625c8af4d6802/examples/opt_serving/model/wrapper.py#L313
2. Run
  ```
  python3 benchmark_text_gen.py --model alpa/opt-2.7b  --debug --num-beams 6
  ```

## Temporary workaround
I did some investigation and found the `scatter-gather/local-allgather` optimization is the root cause.
If I turn this off, the results become correct.
https://github.com/alpa-projects/alpa/blob/b8c4329f403621913e84a202086625c8af4d6802/alpa/global_env.py#L57

**Comment by ZYHowell** [comment]
This is actually a bug for local-allgather at any case that the first dimension is not divisible by num devices. A minimal reproducable case is the `test_4gpu_allgather` in unit test `test_cross_mesh_resharding.py`, but changing the first dimension from 2 to 3. I'm fixing it now.

**Comment by merrymercy** [comment]
Fixed by #590
